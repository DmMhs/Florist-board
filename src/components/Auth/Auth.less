.Auth {
  min-height: calc(100vh - 7rem);
  form {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    width: 50%;
    margin: 0 auto;
    border: 1px solid #ccc;
    border-radius: 8px;
    padding: 1rem;
    .authenticated-message {
      color: #333333;
      h2 {
        font-weight: normal;
        i {
          font-size: 2.5rem;
        }
      }
      span {
        font-size: 1.3rem;
      }
    }
    .form-field {
      width: calc(100% - 2rem);
      margin: 0 1rem;
      font-size: 1.3rem;
      display: flex;
      flex-direction: row;
      justify-content: flex-start;
      flex-wrap: wrap;
      align-items: center;
      padding: 0.7rem;
      label {
        width: 100%;
        text-align: left;
      }
      input {
        width: 50%;
        font-size: 1.2rem;
        padding: 5px;
        border-radius: 5px;
        border: 1px solid #ccc;
      }
    }
    button {
      border: none;
      border-radius: 5px;
      padding: 7px;
      background-color: rgb(98, 182, 103);
      font-size: 1.3rem;
      font-weight: lighter;
      color: white;
      cursor: pointer;
      &:hover {
        background-color: rgb(78, 145, 83);
      }
      &[disabled] {
        background-color: rgb(185, 212, 185);
        pointer-events: none;
      }
    }
    .Popup {
      margin: 0 8px;
      padding: 5px;
      i {
        margin: 0;
        margin-right: 5px;
      }
      i,
      span {
        font-size: 1.1rem;
      }
    }
    hr {
      border: 0.5px solid #eee;
      width: 90%;
    }
    h3 {
      font-weight: normal;
      color: #333333;
    }
    p {
      width: 100%;
      text-align: left;
    }
    .switch-btn {
      background-color: skyblue;
      &:hover {
        background-color: rgb(105, 163, 185);
      }
      a {
        text-decoration: none;
        color: white;
      }
    }
    .google-pic,
    .facebook-pic {
      max-width: 2rem;
      margin: 0 10px;
      cursor: pointer;
      transition: 1s;
      &:hover {
        position: relative;
        bottom: 1px;
        left: 1px;
        transition: 1s;
      }
    }
  }
}

@media screen and (max-width: 768px) {
  .Auth {
    width: 100%;
    form {
      width: 80%;
      margin-top: 4rem;
      margin-bottom: 1rem;
      .form-field {
        font-size: 1.1rem;
        input {
          width: 95%;
        }
      }
    }
    button {
      font-size: 1.1rem;
      margin: 0 auto;
      margin-bottom: 5px;
    }
    .Popup {
      padding: 0px;
      width: fit-content;
    }
  }
}
