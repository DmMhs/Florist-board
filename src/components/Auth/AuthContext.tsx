import React, { Component } from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';

export interface AuthContextState {
  userLogin: string | null | undefined;
  userId: string | null | undefined;
  userToken: string | null | undefined;
  userAuthenticated: boolean;
  authenticationMethod: string | null | undefined;
  lang: string | null | undefined;
}

export const AuthContext = React.createContext({
  state: {} as AuthContextState,
  setUserCredentials: (
    userLogin: string | null,
    userId: string | null,
    userToken: string | null,
    authenticationMethod?: string | null,
    event?: Event
  ) => {},
  logout: () => {},
  setLang: (lang: string | null) => {}
});
class AuthContextProvider extends Component<
  RouteComponentProps<{}>,
  AuthContextState
> {
  constructor(props: RouteComponentProps<{}>) {
    super(props);
    this.state = {
      userLogin: '',
      userId: '',
      userToken: '',
      userAuthenticated: false,
      authenticationMethod: undefined,
      lang: 'en'
    };
  }
  componentDidMount() {
    if (
      localStorage.floristAuthToken === '' ||
      localStorage.floristAuthToken === undefined
    ) {
      this.setState({
        userLogin: '',
        userId: '',
        userToken: '',
        authenticationMethod: '',
        userAuthenticated: false
      });
    } else {
      this.setState({
        userLogin: localStorage.floristAuthLogin,
        userId: localStorage.floristAuthUserId,
        userToken: localStorage.floristAuthToken,
        authenticationMethod: localStorage.floristAuthMethod,
        userAuthenticated: true
      });
    }
  }
  setUserCredentialsHandler = (
    userLogin: string | null | undefined,
    userId: string | null | undefined,
    userToken: string | null | undefined,
    authenticationMethod?: string | null | undefined,
    event?: Event
  ) => {
    this.setState({
      userAuthenticated: true,
      userLogin,
      userId,
      userToken,
      authenticationMethod
    });
    localStorage.floristAuthLogin = userLogin;
    localStorage.floristAuthToken = userToken;
    localStorage.floristAuthUserId = userId;
    localStorage.floristAuthMethod = authenticationMethod;
  };
  logoutHandler = () => {
    this.setState({
      userAuthenticated: false,
      userLogin: '',
      userId: '',
      userToken: '',
      authenticationMethod: undefined
    });
    localStorage.floristAuthLogin = '';
    localStorage.floristAuthToken = '';
    localStorage.floristAuthUserId = '';
    localStorage.floristAuthMethod = '';
  };

  setLangHandler = (lang: string | null) => {
    this.setState({
      lang: lang
    });
  };

  render() {
    return (
      <AuthContext.Provider
        value={{
          state: this.state,
          setUserCredentials: this.setUserCredentialsHandler,
          logout: this.logoutHandler,
          setLang: this.setLangHandler
        }}
      >
        {this.props.children}
      </AuthContext.Provider>
    );
  }
}

export default withRouter<RouteComponentProps<{}>>(AuthContextProvider);
