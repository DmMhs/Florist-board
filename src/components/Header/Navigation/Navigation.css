.Navigation ul {
    list-style: none;
}

.Navigation li {
    display: inline-block;
    margin: 5px;
}

.Navigation a {
    font-size: 1.3rem;
    text-decoration: none;
    color: #333333;
    transition: 0.3s;
}

.Navigation a:hover, 
.Navigation a:active {
    color: tomato;
    transition: 0.3s;
}

.Navigation .active {
    color: tomato;
}