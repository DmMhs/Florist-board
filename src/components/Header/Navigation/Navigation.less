.Navigation {
  ul {
    list-style: none;

    li {
      display: inline-block;
      margin: 5px;

      a {
        margin: 0;
        font-size: 1.3rem;
        text-decoration: none;
        border-radius: 5px;
        padding: 6px;
        background-color: #eee;
        color: #333333;
        cursor: pointer;

        &:hover,
        &:active,
        &.active {
          background-color: tomato;
          color: white;
          transition: 0.3s;
        }
      }

      &.accountLink {
        box-sizing: border-box;
        position: relative;
        background-color: #eee;

        a {
          &.active,
          &:hover,
          &:active {
            background-color: #eee;
            color: #333333;
          }

          &.signup-link:hover,
          &.signup-link:active,
          &.signup-link:active,
          &.signin-link:hover,
          &.signin-link:active,
          &.signin-link:active,
          &.log-out:hover,
          &.log-out:active,
          &.log-out:active {
            background-color: tomato;
            color: white;
          }
        }
      }
    }
  }

  .dropbtn {
    background-color: #eee;
    color: #333333;
    font-size: 1.3rem;
    border-radius: 5px;
    cursor: pointer;
  }

  .dropdown-content {
    display: none;
    position: absolute;
    width: 100%;
    overflow: auto;
    z-index: 1;
    box-sizing: border-box;

    .user-login {
      background-color: #eee;
      color: #333333;
      text-align: center;
      margin: 0;
      width: 100%;
      height: 100%;
    }
    .log-out {
      color: #333333;
    }

    a {
      background-color: #eee;
      border-radius: 0;
      color: #333333;
      padding: 6px;
      text-decoration: none;
      display: block;

      &:last-child {
        border-bottom-left-radius: 5px;
        border-bottom-right-radius: 5px;
      }
    }
  }
  .langLink {
    position: relative;
    .lang-btn {
      display: flex;
      flex-direction: row;
      justify-content: flex-start;
      align-items: center;
      flex-wrap: nowrap;
      padding: 6px;
      span {
        width: 100px;
      }
      &.active {
        background-color: #eee;
        color: #333333;
        border-bottom-left-radius: 0;
        border-bottom-right-radius: 0;
      }
      img {
        max-width: 1.3rem;
        margin: 0 5px;
      }
      i {
        font-size: 1.3rem;
      }
    }
    .lang-options {
      display: none;
      position: absolute;
      width: 100%;
      overflow: auto;
      z-index: 1;
      background-color: #eee;
      border-radius: 0;
      border-bottom-left-radius: 5px;
      border-bottom-right-radius: 5px;
      a {
        width: 100%;
        display: flex;
        flex-direction: row;
        align-items: center;
        justify-content: flex-start;
        box-sizing: border-box;
        span {
          width: 100px;
        }
      }
      &.show {
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
      }
      &:hover {
        cursor: pointer;
        background-color: tomato;
      }
      img {
        max-width: 1.3rem;
        margin: 0 5px;
      }
    }
  }

  .show {
    display: block;
    z-index: 1;
  }
}
@media screen and (max-width: 1050px) {
  .Navigation {
    ul {
      li {
        a {
          font-size: 1rem;
        }
        &.accountLink {
          a {
            font-size: 1rem;
          }
          .dropdown-content {
            font-size: 0.8rem;
          }
        }
      }
    }
    .langLink {
      position: fixed;
      bottom: 3rem;
      margin: 10 20px;
      width: 100%;
      .lang-btn {
        span {
          width: 50px;
        }
      }
      .lang-options {
        margin: 0 auto;
        margin-top: 2rem;
        width: calc(75% + 12px);
      }
    }
  }
}
@media screen and (max-width: 768px) {
  .Navigation {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: white;
    z-index: 5;
    opacity: 0.95;
    ul {
      margin-top: 4rem;
      display: flex;
      flex-direction: column;
      justify-content: flex-start;
      align-items: flex-start;
      margin: 0;
      padding: 0;
      margin-top: 4rem;
      li {
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        align-items: center;
        width: 100%;
        a {
          display: inline-block;
          margin: 0 auto;
          width: 75%;
          font-size: 1.3rem;
        }
        &.accountLink {
          box-sizing: content-box;
          border-radius: 5px;
          width: fit-content;
          position: relative;
          margin: 0 auto;
          margin-top: 5px;
          width: 75%;
          a {
            font-size: 1.3rem;
          }
        }
        &.langLink {
          position: fixed;
          bottom: 3rem;
          margin: 10 20px;
          width: 100%;
          .lang-btn {
            span {
              width: 70px;
            }
          }
          .lang-options {
            margin: 0 auto;
            margin-top: 2rem;
            width: calc(75% + 12px);
          }
        }
      }
    }
    .dropbtn {
      width: fit-content;
      background-color: #eee;
      white-space: nowrap;
      color: #333333;
      padding: 6px;
      border-radius: 5px;
      cursor: pointer;
      width: 100%;
    }
    .dropdown-content {
      display: none;
      position: absolute;
      padding: 0;
      left: 0;
      right: 0;
      width: calc(100% + 12px);
      div {
        width: calc(100% - 12px);
        a {
          padding: 6px;
        }
      }
      .log-out {
        display: inline-block;
        width: calc(100% - 12px);
        color: darkred;
      }
      .signin-link,
      .signup-link {
        width: 100%;
      }
    }
    .langLink {
      position: fixed;
      bottom: 3rem;
      margin: 10 20px;
      width: 100%;
      .lang-btn {
        justify-content: center;
        span {
          width: 100%;
        }
      }
      .lang-options {
        margin: 0 auto;
        margin-top: 2rem;
        width: calc(75% + 12px);
        a {
          justify-content: center;
        }
      }
    }
    .show {
      display: block;
      margin-top: 2rem;
      z-index: 1;
    }
  }
}
