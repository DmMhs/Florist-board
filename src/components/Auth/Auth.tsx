import React, { Component } from 'react';
import firebase from 'firebase';
import axios from 'axios';
import {
  withRouter,
  RouteComponentProps as RCProps,
  NavLink
} from 'react-router-dom';

import Popup from '../Popup/Popup';
import { AppContext } from '../../AppContext';
import labels from '../../config/labels';
import './Auth.less';

interface MatchParams {
  mode: string;
}

interface Props extends RouteComponentProps<MatchParams> {}

interface RouteComponentProps<P> {
  match: match<P>;
}

interface match<P> {
  params: P;
  isExact: boolean;
  path: string;
  url: string;
}

interface AuthState {
  formData: {
    email: string;
    password: string;
  };
  mode: string;
}

class Auth extends Component<
  RouteComponentProps<MatchParams> & RCProps<{}>,
  AuthState
> {
  public static getDerivedStateFromProps(
    props: RouteComponentProps<MatchParams> & RCProps<{}>,
    state: AuthState
  ) {
    return {
      mode: props.match.params.mode
    };
  }

  constructor(props: RouteComponentProps<MatchParams> & RCProps<{}>) {
    super(props);
    this.state = {
      formData: {
        email: '',
        password: ''
      },
      mode: this.props.match.params.mode
    };
  }

  private emailInputChangeHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedFormData = { ...this.state.formData };
    updatedFormData.email = event.target.value;
    this.setState({
      formData: updatedFormData
    });
  };

  private passwordInputChangeHandler = (
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedFormData = { ...this.state.formData };
    updatedFormData.password = event.target.value;
    this.setState({
      formData: updatedFormData
    });
  };

  private getIdToken = () => {
    return firebase
      .auth()
      .currentUser!.getIdToken(true)
      .then(idToken => {
        return idToken;
      })
      .catch(error => {
        console.log(error);
      });
  };

  private formSubmitHandler = (
    event: React.MouseEvent<HTMLFormElement, MouseEvent>
  ) => {
    event.preventDefault();
    const value = this.context;
    const initialFormData = {
      email: '',
      password: ''
    };
    const authData = {
      email: this.state.formData.email,
      password: this.state.formData.password,
      returnSecureToken: true
    };

    if (this.state.mode === 'signup') {
      this.setState({
        formData: initialFormData
      });
      firebase
        .auth()
        .createUserWithEmailAndPassword(authData.email, authData.password)
        .then(response => {
          (this.props as RouteComponentProps<MatchParams> &
            RCProps<{}>).history.push('/auth/signin');
        })
        .catch(error => {
          console.log(error);
        });
    } else {
      this.setState({
        formData: initialFormData
      });
      firebase
        .auth()
        .signInWithEmailAndPassword(authData.email, authData.password)
        .then(async response => {
          const idToken = await this.getIdToken();
          value.setUserCredentials(
            response.user!.email,
            response.user!.uid,
            idToken
          );
          (this.props as RouteComponentProps<MatchParams> &
            RCProps<{}>).history.push('/');
        })
        .catch(error => console.log(error));
    }
  };

  private authWithGoogleHandler = () => {
    const value = this.context;
    firebase
      .auth()
      .signInWithPopup(new firebase.auth.GoogleAuthProvider())
      .then(async response => {
        const idToken = await this.getIdToken();
        value.setUserCredentials(
          response.user!.email,
          response.user!.uid,
          idToken,
          'google'
        );
        await axios.post(
          'https://us-central1-florist-cb933.cloudfunctions.net/giveAdminRole',
          {
            admin: true
          }
        );
        (this.props as RouteComponentProps<MatchParams> &
          RCProps<{}>).history.push('/');
      })
      .catch(error => console.log(error));
  };

  private authWithFacebookHandler = () => {
    const value = this.context;
    firebase
      .auth()
      .signInWithPopup(new firebase.auth.FacebookAuthProvider())
      .then(async response => {
        const idToken = await this.getIdToken();
        value.setUserCredentials(
          response.user!.displayName,
          response.user!.uid,
          idToken,
          'facebook'
        );
        (this.props as RouteComponentProps<MatchParams> &
          RCProps<{}>).history.push('/');
      })
      .catch(error => console.log(error));
  };

  public render() {
    const { email, password } = this.state.formData;
    const { mode } = this.state;
    const context = this.context;

    const infoPopup = (
      <Popup
        type="info"
        message={
          context.state.lang === 'en'
            ? 'Please, provide a required data'
            : 'Будь ласка, введіть дані'
        }
      />
    );
    return (
      <div className="Auth">
        <AppContext.Consumer>
          {value =>
            value && (
              <form
                onSubmit={
                  this.formSubmitHandler as
                    | ((event: React.FormEvent<HTMLFormElement>) => void)
                    | undefined
                }
              >
                {value.state.userAuthenticated === false ? (
                  <div>
                    <div className="form-field">
                      <label>
                        {labels[value.state.lang as string].pages.auth.email}
                      </label>
                      <input
                        type="email"
                        className="email-input"
                        onChange={this.emailInputChangeHandler}
                        value={email}
                        required
                      />
                    </div>
                    <div className="form-field">
                      <label>
                        {labels[value.state.lang as string].pages.auth.password}
                      </label>
                      <input
                        type="password"
                        className="password-input"
                        minLength={6}
                        onChange={this.passwordInputChangeHandler}
                        value={password}
                        required
                      />
                    </div>
                    <div className="form-field">
                      <button
                        type="submit"
                        disabled={
                          password === '' || email === '' ? true : false
                        }
                        className="submit-btn"
                      >
                        {labels[value.state.lang as string].pages.auth.btn}
                      </button>
                      {password === '' || email === '' ? infoPopup : null}
                    </div>
                    <hr />
                    <h3>
                      {
                        labels[value.state.lang as string].pages.auth
                          .alternative
                      }
                    </h3>
                    <div className="form-field">
                      <p>
                        {mode === 'signup'
                          ? labels[value.state.lang as string].pages.auth.signup
                              .google
                          : labels[value.state.lang as string].pages.auth.signin
                              .google}
                      </p>{' '}
                      <br />
                      <a
                        onClick={this.authWithGoogleHandler}
                        className="google-auth"
                      >
                        <img
                          src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Fgoogle.png?alt=media&token=3210d61d-cad2-45bc-a343-0ed08c097bb6"
                          alt="google-pic"
                          className="google-pic"
                        />
                      </a>
                    </div>
                    <div className="form-field">
                      <p>
                        {mode === 'signup'
                          ? labels[value.state.lang as string].pages.auth.signup
                              .facebook
                          : labels[value.state.lang as string].pages.auth.signin
                              .facebook}
                      </p>
                      <br />
                      <a
                        onClick={this.authWithFacebookHandler}
                        className="facebook-auth"
                      >
                        <img
                          src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Ffacebook.png?alt=media&token=32a65be5-5336-45fd-8a50-44f8880ddbd0"
                          alt="facebook-pic"
                          className="facebook-pic"
                        />
                      </a>
                    </div>
                    <hr />
                    <div className="form-field">
                      <p>
                        {mode === 'signup'
                          ? labels[value.state.lang as string].pages.auth.signup
                              .account
                          : labels[value.state.lang as string].pages.auth.signin
                              .account}
                      </p>
                      <button type="button" className="switch-btn">
                        <NavLink
                          to={`/auth/${
                            mode === 'signup' ? 'signin' : 'signup'
                          }`}
                        >
                          {mode === 'signup'
                            ? labels[value.state.lang as string].pages.auth
                                .signup.btn
                            : labels[value.state.lang as string].pages.auth
                                .signin.btn}
                        </NavLink>
                      </button>
                    </div>
                  </div>
                ) : (
                  <div className="authenticated-message">
                    {' '}
                    <h2>
                      {
                        labels[value.state.lang as string].pages.auth.isAuth
                          .main
                      }{' '}
                      <i className="far fa-flushed" />
                    </h2>{' '}
                    <span>
                      {labels[value.state.lang as string].pages.auth.isAuth.sub}
                    </span>
                  </div>
                )}
              </form>
            )
          }
        </AppContext.Consumer>
      </div>
    );
  }
}

Auth.contextType = AppContext;
export default withRouter<RouteComponentProps<MatchParams> & RCProps<{}>>(Auth);
