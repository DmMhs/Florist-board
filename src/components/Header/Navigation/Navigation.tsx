import React, { Component, RefObject } from 'react';
import { NavLink, withRouter, RouteComponentProps } from 'react-router-dom';
import firebase from 'firebase';

import { AppContext } from '../../../AppContext';
import labels from '../../../config/labels';
import './Navigation.less';

interface NavigationState {}

class Navigation extends Component<RouteComponentProps<{}>, NavigationState> {
  private authOptionsToggleRef: RefObject<HTMLAnchorElement>;
  private authOptionsRef: RefObject<HTMLDivElement>;
  private langOptionsRef: RefObject<HTMLDivElement>;
  private langOptionsToggleRef: RefObject<HTMLAnchorElement>;

  constructor(props: RouteComponentProps<{}>) {
    super(props);
    this.authOptionsRef = React.createRef();
    this.authOptionsToggleRef = React.createRef();
    this.langOptionsRef = React.createRef();
    this.langOptionsToggleRef = React.createRef();
  }

  private accountClickedHandler = () => {
    this.authOptionsToggleRef.current!.classList.toggle('active');
    this.authOptionsRef.current!.classList.toggle('show');
  };

  private logoutClickedHandler = () => {
    const value = this.context;
    firebase.auth().signOut();
    value.logout();
    if ((this.props as RouteComponentProps<{}>).history !== undefined) {
      (this.props as RouteComponentProps<{}>).history.push('/');
    }
  };

  private langBtnClickedHandler = () => {
    this.langOptionsRef.current!.classList.toggle('show');
    this.langOptionsToggleRef.current!.classList.toggle('active');
  };
  private langOptionClickedHandler = (option: string) => {
    const context = this.context;
    context.setLang(option);
  };
  public render() {
    return (
      <div className="Navigation">
        <AppContext.Consumer>
          {value =>
            value && (
              <ul>
                <li>
                  <NavLink to="/" exact={true}>
                    {labels[value.state.lang as string].navigation.home}{' '}
                    <i className="fas fa-home" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/shop">
                    {labels[value.state.lang as string].navigation.shop}{' '}
                    <i className="fas fa-leaf" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/gallery">
                    {labels[value.state.lang as string].navigation.gallery}{' '}
                    <i className="far fa-image" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/contacts">
                    {labels[value.state.lang as string].navigation.contacts}{' '}
                    <i className="fas fa-map-marker-alt" />
                  </NavLink>
                </li>
                <li className="accountLink">
                  <a
                    className="dropbtn"
                    onClick={this.accountClickedHandler}
                    ref={this.authOptionsToggleRef}
                  >
                    {labels[value.state.lang as string].navigation.account.main}{' '}
                    <i className="fas fa-user-circle" />{' '}
                    <i className="fas fa-caret-down" />
                  </a>
                  <div className="dropdown-content" ref={this.authOptionsRef}>
                    {value.state.userLogin!.length > 0 ? (
                      <p className="user-login">{value.state.userLogin}</p>
                    ) : null}
                    <NavLink to="/admin">
                      <i className="fas fa-users-cog" />
                    </NavLink>
                    {value.state.userAuthenticated === false ? (
                      <div>
                        <NavLink to="/auth/signup" className="signup-link">
                          {
                            labels[value.state.lang as string].navigation
                              .account.menu.signUp
                          }{' '}
                          <i className="fas fa-user-plus" />
                        </NavLink>
                        <NavLink to="/auth/signin" className="signin-link">
                          {
                            labels[value.state.lang as string].navigation
                              .account.menu.signIn
                          }{' '}
                          <i className="fas fa-sign-in-alt" />
                        </NavLink>
                      </div>
                    ) : (
                      <a
                        onClick={this.logoutClickedHandler}
                        className="log-out"
                      >
                        {
                          labels[value.state.lang as string].navigation.account
                            .menu.logOut
                        }
                      </a>
                    )}
                  </div>
                </li>
                <li className="langLink">
                  <a
                    className="lang-btn"
                    onClick={this.langBtnClickedHandler}
                    ref={this.langOptionsToggleRef}
                  >
                    <span>
                      {labels[value.state.lang as string].navigation.lang}
                    </span>
                    {value.state.lang === 'en' ? (
                      <img src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Flang%2Funited-kingdom.png?alt=media&token=791aa1e1-7dc6-467f-a716-3eb8dce0b313" />
                    ) : (
                      <img src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Flang%2Fukraine.png?alt=media&token=9305aa3b-0e62-411e-b6ae-7dbbcaa72e74" />
                    )}{' '}
                    <i className="fas fa-caret-down" />
                  </a>
                  <div className="lang-options" ref={this.langOptionsRef}>
                    {value.state.lang === 'uk' ? (
                      <a
                        onClick={this.langOptionClickedHandler.bind(this, 'en')}
                      >
                        <span>{labels.en.navigation.lang}</span>{' '}
                        <img src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Flang%2Funited-kingdom.png?alt=media&token=791aa1e1-7dc6-467f-a716-3eb8dce0b313" />
                      </a>
                    ) : (
                      <a
                        onClick={this.langOptionClickedHandler.bind(this, 'uk')}
                      >
                        <span>{labels.uk.navigation.lang}</span>
                        <img src="https://firebasestorage.googleapis.com/v0/b/florist-cb933.appspot.com/o/icons%2Flang%2Fukraine.png?alt=media&token=9305aa3b-0e62-411e-b6ae-7dbbcaa72e74" />
                      </a>
                    )}
                  </div>
                </li>
              </ul>
            )
          }
        </AppContext.Consumer>
      </div>
    );
  }
}

Navigation.contextType = AppContext;
export default withRouter<RouteComponentProps<{}>>(Navigation);
