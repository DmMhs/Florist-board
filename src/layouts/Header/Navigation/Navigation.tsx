import React, { Component, RefObject } from 'react';
import { NavLink, withRouter, RouteComponentProps } from 'react-router-dom';
import firebase from 'firebase';

import { AppContext } from '../../../AppContext';
import { urls } from '../../../config/urls';
import './Navigation.less';

interface NavigationState {}

class Navigation extends Component<RouteComponentProps<{}>, NavigationState> {
  private authOptionsToggleRef: RefObject<HTMLAnchorElement>;
  private authOptionsRef: RefObject<HTMLDivElement>;
  private langOptionsRef: RefObject<HTMLDivElement>;
  private langOptionsToggleRef: RefObject<HTMLAnchorElement>;

  constructor(props: RouteComponentProps<{}>) {
    super(props);
    this.authOptionsRef = React.createRef();
    this.authOptionsToggleRef = React.createRef();
    this.langOptionsRef = React.createRef();
    this.langOptionsToggleRef = React.createRef();
  }

  private accountClickedHandler = () => {
    this.authOptionsToggleRef.current!.classList.toggle('active');
    this.authOptionsRef.current!.classList.toggle('show');
  };

  private logoutClickedHandler = () => {
    const value = this.context;
    firebase.auth().signOut();
    value.logout();
    if ((this.props as RouteComponentProps<{}>).history !== undefined) {
      (this.props as RouteComponentProps<{}>).history.push('/');
    }
  };

  private langBtnClickedHandler = () => {
    this.langOptionsRef.current!.classList.toggle('show');
    this.langOptionsToggleRef.current!.classList.toggle('active');
  };
  private langOptionClickedHandler = (option: string) => {
    const context = this.context;
    context.setLang(option);
  };
  public render() {
    const context = this.context;
    const lang = context.state.lang;
    const labels = context.state.labels;
    const labelsRoot = labels[lang as string].navigation;

    return (
      <div className="Navigation">
        <AppContext.Consumer>
          {value =>
            value && (
              <ul>
                <li>
                  <NavLink to="/" exact={true}>
                    {labelsRoot.home} <i className="fas fa-home" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/shop">
                    {labelsRoot.shop} <i className="fas fa-leaf" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/gallery">
                    {labelsRoot.gallery} <i className="far fa-image" />
                  </NavLink>
                </li>
                <li>
                  <NavLink to="/contacts">
                    {labelsRoot.contacts}{' '}
                    <i className="fas fa-map-marker-alt" />
                  </NavLink>
                </li>
                <li className="accountLink">
                  <a
                    className="dropbtn"
                    onClick={this.accountClickedHandler}
                    ref={this.authOptionsToggleRef}
                  >
                    {labelsRoot.account.main}{' '}
                    <i className="fas fa-user-circle" />{' '}
                    <i className="fas fa-caret-down" />
                  </a>

                  <div className="dropdown-content" ref={this.authOptionsRef}>
                    {value.state.userLogin!.length > 0 ? (
                      <p className="user-login">{value.state.userLogin}</p>
                    ) : null}

                    {value.state.userRole === 'admin' ? (
                      <div>
                        <NavLink to="/admin/add-product" className="admin">
                          <i className="fas fa-users-cog" />
                        </NavLink>
                      </div>
                    ) : null}

                    {value.state.userAuthenticated === false ? (
                      <div>
                        <NavLink to="/auth/signup" className="signup-link">
                          {labelsRoot.account.menu.signUp}{' '}
                          <i className="fas fa-user-plus" />
                        </NavLink>
                        <NavLink to="/auth/signin" className="signin-link">
                          {labelsRoot.account.menu.signIn}{' '}
                          <i className="fas fa-sign-in-alt" />
                        </NavLink>
                      </div>
                    ) : (
                      <a
                        onClick={this.logoutClickedHandler}
                        className="log-out"
                      >
                        {labelsRoot.account.menu.logOut}
                      </a>
                    )}
                  </div>
                </li>
                <li className="langLink">
                  <a
                    className="lang-btn"
                    onClick={this.langBtnClickedHandler}
                    ref={this.langOptionsToggleRef}
                  >
                    <span>{labelsRoot.lang}</span>
                    {value.state.lang === 'en' ? (
                      <img src={urls.en_flag} />
                    ) : (
                      <img src={urls.ua_flag} />
                    )}{' '}
                    <i className="fas fa-caret-down" />
                  </a>
                  <div className="lang-options" ref={this.langOptionsRef}>
                    {value.state.lang === 'ua' ? (
                      <a
                        onClick={this.langOptionClickedHandler.bind(this, 'en')}
                      >
                        <span>{labels.en.navigation.lang}</span>{' '}
                        <img src={urls.en_flag} />
                      </a>
                    ) : (
                      <a
                        onClick={this.langOptionClickedHandler.bind(this, 'ua')}
                      >
                        <span>{labels.ua.navigation.lang}</span>
                        <img src={urls.ua_flag} />
                      </a>
                    )}
                  </div>
                </li>
              </ul>
            )
          }
        </AppContext.Consumer>
      </div>
    );
  }
}

Navigation.contextType = AppContext;
export default withRouter<RouteComponentProps<{}>>(Navigation);
