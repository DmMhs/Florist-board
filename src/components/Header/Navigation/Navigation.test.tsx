import React from 'react';
import { shallow, mount } from 'enzyme';
import PropTypes from 'prop-types';

import Navigation from './Navigation';
import { BrowserRouter } from 'react-router-dom';
import { AuthContext } from '../../Auth/AuthContext';

it('Navigation component matches a snapshot', () => {
  const wrapper = shallow(<Navigation />);
  expect(wrapper).toMatchSnapshot();
});
