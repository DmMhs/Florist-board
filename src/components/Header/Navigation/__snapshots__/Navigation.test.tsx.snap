// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation component matches a snapshot 1`] = `
<Route>
  <Component />
</Route>
`;
